MagneticField/GeomBuilder/plugins/dd4hep/volumeHandle.cc

```diff
@@ -0,0 +1,208 @@
+  //
+  // For pseudotraps the refPlane is parallel to beam line and global R is
+  // along Z, global phi is along +-X and and global Z along Y:
+  //
+  //   Global(for vol at pi/2)    Local
+  //   +R (+Y)                    +Z
+  //   +phi(-X)                   +X
+  //   +Z                         +Y
+  //
+  // Note that the frame is centered in the DDD volume center, which is
+  // inside the volume for DDD boxes and (pesudo)trapezoids, on the beam line
+  // for tubs, cons and trunctubs.
+
+  // In geometry version 1103l, trapezoids have X and Z in the opposite direction
+  // than the above.  Boxes are either oriented as described above or in some case
+  // have opposite direction for Y and X.
+
+  // The global position
+  Surface::PositionType &posResult = center_;
+
+  // The reference plane rotation
+  math::XYZVector x, y, z;
+  dd4hep::Rotation3D refRot;
+  fv.rot(refRot);
+  refRot.GetComponents(x, y, z);
+  if (debug) {
+    if (x.Cross(y).Dot(z) < 0.5) {
+      LogDebug("magneticfield::volumeHandle") << "*** WARNING: Rotation is not RH ";
+    }
+  }
+
+  // The global rotation
+  Surface::RotationType rotResult(float(x.X()),
+                                  float(x.Y()),
+                                  float(x.Z()),
+                                  float(y.X()),
+                                  float(y.Y()),
+                                  float(y.Z()),
+                                  float(z.X()),
+                                  float(z.Y()),
+                                  float(z.Z()));
+
+  refPlane = new GloballyPositioned<float>(posResult, rotResult);
+
+  // Check correct orientation
+  if (debug) {
+    LogDebug("magneticfield::volumeHandle") << "Refplane pos  " << refPlane->position();
+
+    // See comments above for the conventions for orientation.
+    LocalVector globalZdir(0., 0., 1.);  // Local direction of the axis along global Z
+
+    /* Preserve in case pseudotrap is needed again
+    if (shape_ == DDSolidShape::ddpseudotrap) {
+      globalZdir = LocalVector(0., 1., 0.);
+    }
+    */
+    if (refPlane->toGlobal(globalZdir).z() < 0.) {
+      globalZdir = -globalZdir;
+    }
+    float chk = refPlane->toGlobal(globalZdir).dot(GlobalVector(0, 0, 1));
+    if (chk < .999)
+      LogDebug("magneticfield::volumeHandle") << "*** WARNING RefPlane check failed!***" << chk;
+  }
+}
+
+std::vector<VolumeSide> volumeHandle::sides() const {
+  std::vector<VolumeSide> result;
+  for (int i = 0; i < 6; ++i) {
+    // If this is just a master volume out of wich a 2pi volume
+    // should be built (e.g. central cylinder), skip the phi boundaries.
+    if (expand && (i == phiplus || i == phiminus))
+      continue;
+
+    // FIXME: Skip null inner degenerate cylindrical surface
+    if (shape_ == DDSolidShape::ddtubs && i == SurfaceOrientation::inner && theRMin < 0.001)
+      continue;
+
+    ReferenceCountingPointer<Surface> s = const_cast<Surface *>(surfaces[i].get());
```

Comment:
can this const_cast be avoided?
this triggers a static analyzer issue
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2856d3/2029/llvm-analysis/report-a42427.html#EndPath

Reply:
The `VolumeSide` constructor requires a non-const pointer, so I changed `surfaces` to the non-const `ReferenceCountingPointer<Surface>`. An alternative is to change the `VolumeSide` class, but that would affect even more packages.